impeller/core/sampler.h

```diff
@@ -31,8 +31,19 @@ class Sampler {
   Sampler& operator=(const Sampler&) = delete;
 };
 
+/// @brief A sampler subclass to be returned in the result of a lost device
+//         or other failure to allocate a sampler object.
+class InvalidSampler : public Sampler {
```

Comment:
I'm not a huge fan of creating an object to represent what essentially a null value.  I would prefer that instead of passing in a `const Sampler& sampler` we passed in `const std::unique_ptr<Sampler>& sampler`.  Then we can keep the semantics all the same.

I think `IsValid()` is an anti-pattern.  We shouldn't have invalid objects.  It starts polluting your code where you have to check for validity all the time.  Or it doesn't and you just have potential bugs everywhere.

Reply:
You already are switching to a unique_ptr but you are dereferencing it below.  All you have to do is not dereference this and pass the reference to the unique_ptr.